### PR TITLE
[Test-Proxy] Remove Explicit List of Content-Headers

### PR DESCRIPTION
**Pull request by scbedd**
@JoshLove-msft @mikeharder 

Tested locally with record/playback of a few tables tests. Unfortunately that highlighted an issue with my previous PR during my local testing.

### CODE REVIEW

File: tools/test-proxy/Azure.Sdk.Tools.TestProxy/RecordingHandler.cs [hunk 1]
```diff
@@ -50,6 +50,8 @@ public RecordingHandler(string targetDirectory)
         private static readonly string[] s_contentRequestHeaders = new string[] {
```
Comment (JoshLove-msft) [inline tools/test-proxy/Azure.Sdk.Tools.TestProxy/RecordingHandler.cs:50]: I think this was meant to be removed?

Comment (JoshLove-msft) [reply tools/test-proxy/Azure.Sdk.Tools.TestProxy/RecordingHandler.cs]: Nevermind!

Comment (scbedd) [reply tools/test-proxy/Azure.Sdk.Tools.TestProxy/RecordingHandler.cs]: Yeah, commit just wasn't included. Fixed now.

File: tools/test-proxy/Azure.Sdk.Tools.TestProxy/RecordingHandler.cs [hunk 2]
```diff
@@ -276,29 +270,21 @@ private HttpRequestMessage CreateUpstreamRequest(HttpRequest incomingRequest, by
...
-                    else
+
+                    if(!upstreamRequest.Content.Headers.TryAddWithoutValidation(header.Key, values))
                     {
-                        if (!header.Key.StartsWith("x-recording"))
-                        {
-                            upstreamRequest.Headers.TryAddWithoutValidation(header.Key, values);
-                        }
+                        throw new HttpException(
+                            HttpStatusCode.InternalServerError,
```
Comment (JoshLove-msft) [inline tools/test-proxy/Azure.Sdk.Tools.TestProxy/RecordingHandler.cs:283]: Should this be a bad request instead?

**Comment by JoshLove-msft** [comment]
Consider adding a few test cases for this.

**Review by JoshLove-msft — APPROVED** [review]

File: tools/test-proxy/Azure.Sdk.Tools.TestProxy/RecordingHandler.cs [hunk 3]
```diff
@@ -276,29 +270,21 @@ private HttpRequestMessage CreateUpstreamRequest(HttpRequest incomingRequest, by
             {
                 IEnumerable<string> values = header.Value;
 
-                if (s_excludedRequestHeaders.Contains(header.Key, StringComparer.OrdinalIgnoreCase))
```
Comment (JoshLove-msft) [inline tools/test-proxy/Azure.Sdk.Tools.TestProxy/RecordingHandler.cs:279]: Why was this removed?

File: tools/test-proxy/Azure.Sdk.Tools.TestProxy/RecordingHandler.cs (hunk 2 above)
Comment (scbedd) [reply tools/test-proxy/Azure.Sdk.Tools.TestProxy/RecordingHandler.cs:283]: It could be. I was thinking about it from

> We should never hit this situation ever, something has gone massively wrong on the server side.

There should be nothing a client can do that puts us in this state, hence why I'm blaming the server (httpinternal) vs client (badrequest).

Comment (JoshLove-msft) [reply tools/test-proxy/Azure.Sdk.Tools.TestProxy/RecordingHandler.cs:283]: Couldn't a client add an unknown header?

Comment (JoshLove-msft) [reply tools/test-proxy/Azure.Sdk.Tools.TestProxy/RecordingHandler.cs:283]: Or would custom headers get added here?

File: tools/test-proxy/Azure.Sdk.Tools.TestProxy/RecordingHandler.cs (hunk 3 above)
Comment (scbedd) [reply tools/test-proxy/Azure.Sdk.Tools.TestProxy/RecordingHandler.cs:279]: Good point. It would handle dups.

File: tools/test-proxy/Azure.Sdk.Tools.TestProxy/RecordingHandler.cs (hunk 2 above)
Comment (JoshLove-msft) [reply tools/test-proxy/Azure.Sdk.Tools.TestProxy/RecordingHandler.cs:283]: Nevermind - I guess custom headers would still get added so I think it is fine how you have it.

File: tools/test-proxy/Azure.Sdk.Tools.TestProxy/RecordingHandler.cs (hunk 3 above)
Comment (scbedd) [reply tools/test-proxy/Azure.Sdk.Tools.TestProxy/RecordingHandler.cs:279]: Actually. No. I misread this. I definitely don't want this to be here. It DOES need to handle duplicates here though. Otherwise we will tryAdd it, not (because it's already there), and go to add it to content erroneously.

Comment (JoshLove-msft) [reply tools/test-proxy/Azure.Sdk.Tools.TestProxy/RecordingHandler.cs:279]: Not sure I follow - this list contains headers that we don't want to include on the request, so why are we removing the filter? If we should remove it, should we also delete the field?

File: tools/test-proxy/Azure.Sdk.Tools.TestProxy/RecordingHandler.cs (hunk 2 above)
Comment (scbedd) [reply tools/test-proxy/Azure.Sdk.Tools.TestProxy/RecordingHandler.cs:288]: Any unknown custom headers should add just fine to the standard headers dictionary. That's where our `x-recording-Y` go. We should never get to this code. I will just make it a bad request.

File: tools/test-proxy/Azure.Sdk.Tools.TestProxy.Tests/RecordingHandlerTests.cs [hunk 4]
```diff
@@ -23,6 +23,34 @@ namespace Azure.Sdk.Tools.TestProxy.Tests
 {
     public class RecordingHandlerTests
     {
+        private HttpContext _generateHttpRequestContext(string[] headerValueStrings)
```
Comment (JoshLove-msft) [inline tools/test-proxy/Azure.Sdk.Tools.TestProxy.Tests/RecordingHandlerTests.cs:26]:
```suggestion
        private HttpContext GenerateHttpRequestContext(string[] headerValueStrings)
```

File: tools/test-proxy/Azure.Sdk.Tools.TestProxy.Tests/RecordingHandlerTests.cs [hunk 5]
```diff
@@ -23,6 +23,34 @@ namespace Azure.Sdk.Tools.TestProxy.Tests
...
+                context.Request.Headers.TryAdd(hTuple.Item1, hTuple.Item2);
+            }
+
+            context.Request.Headers.TryAdd("x-recording-upstream-base-uri", new string[] { "https://hello-world" });
+            context.Request.Method = "POST";
+
+            return context;
+        }
+
+        private IEnumerable<Tuple<string, StringValues>> _generateHeaderValuesTuples(string[] headerValueStrings) {
```
Comment (JoshLove-msft) [inline tools/test-proxy/Azure.Sdk.Tools.TestProxy.Tests/RecordingHandlerTests.cs:42]:
```suggestion
        private IEnumerable<Tuple<string, StringValues>> GenerateHeaderValuesTuples(string[] headerValueStrings) {
```

File: tools/test-proxy/Azure.Sdk.Tools.TestProxy.Tests/RecordingHandlerTests.cs [hunk 6]
```diff
@@ -23,6 +23,34 @@ namespace Azure.Sdk.Tools.TestProxy.Tests
 {
     public class RecordingHandlerTests
     {
+        private HttpContext GenerateHttpRequestContext(string[] headerValueStrings)
+        {
+            HttpContext context = new DefaultHttpContext();
+
+            foreach(var hTuple in _generateHeaderValuesTuples(headerValueStrings))
```
Comment (JoshLove-msft) [inline tools/test-proxy/Azure.Sdk.Tools.TestProxy.Tests/RecordingHandlerTests.cs:30]:
```suggestion
            foreach(var hTuple in GenerateHeaderValuesTuples(headerValueStrings))
```

File: tools/test-proxy/Azure.Sdk.Tools.TestProxy.Tests/RecordingHandlerTests.cs [hunk 7]
```diff
@@ -564,6 +592,45 @@ public async Task CreateEntryUsesAbsoluteUri()
...
         }
+
+
+        [Theory]
+        [InlineData("Content-Type:application/json; odata=minimalmetadata; streaming=true", "Accept:application/json;odata=minimalmetadata")]
+        [InlineData("Content-MD5:<ContentHash>", "x-ms-version:2019-02-02", "RequestMethod:POST", "Connection:keep-alive")]
+        [InlineData("Content-Encoding:utf-8", "x-ms-version:2019-02-02", "RequestMethod:POST", "Content-Length:50")]
+        public void TestCreateUpstreamRequestIncludesExpectedHeaders(params string[] incomingHeaders)
+        {
+            var requestContext = _generateHttpRequestContext(incomingHeaders);
```
Comment (JoshLove-msft) [inline tools/test-proxy/Azure.Sdk.Tools.TestProxy.Tests/RecordingHandlerTests.cs:603]:
```suggestion
            var requestContext = GenerateHttpRequestContext(incomingHeaders);
```

File: tools/test-proxy/Azure.Sdk.Tools.TestProxy.Tests/RecordingHandlerTests.cs [hunk 8]
```diff
@@ -564,6 +592,45 @@ public async Task CreateEntryUsesAbsoluteUri()
...
+            var recordingHandler = new RecordingHandler(Directory.GetCurrentDirectory());
+            var upstreamRequestContext = _generateHttpRequestContext(incomingHeaders);
```
Comment (JoshLove-msft) [inline tools/test-proxy/Azure.Sdk.Tools.TestProxy.Tests/RecordingHandlerTests.cs:605]:
```suggestion
            var upstreamRequestContext = GenerateHttpRequestContext(incomingHeaders);
```

File: tools/test-proxy/Azure.Sdk.Tools.TestProxy.Tests/RecordingHandlerTests.cs [hunk 9]
```diff
@@ -564,6 +592,45 @@ public async Task CreateEntryUsesAbsoluteUri()
...
+
+            var output = recordingHandler.CreateUpstreamRequest(upstreamRequestContext.Request, new byte[] { });
+
+            // iterate across the set we know about and confirm that GenerateUpstreamRequest worked properly!
+            var setOfHeaders = _generateHeaderValuesTuples(incomingHeaders);
```
Comment (JoshLove-msft) [inline tools/test-proxy/Azure.Sdk.Tools.TestProxy.Tests/RecordingHandlerTests.cs:610]:
```suggestion
            var setOfHeaders = GenerateHeaderValuesTuples(incomingHeaders);
```
